### ISSUE TITLE
expose the canvas context to the visualizer

### ISSUE THREAD

**Issue by katestange** [numberscope/frontscope#118]
The visualizers should have access to native HTML5 canvas context functionality.  As a way of explaining what I mean and why and how one normally does this, I'll go with the example I'm working on for the factoring visualizer, below.

I'm writing a visualizer that calls for a lot of gradient rectangles.  Using the [p5 example](https://p5js.org/examples/color-linear-gradient.html) for this, it is painfully slow.  Instead, you can call p5's [drawingContext](https://p5js.org/reference/#/p5/drawingContext) to get access to the native HTML5 canvas context, which does gradients quickly ([example](https://editor.p5js.org/jonippolito/sketches/iVaT_YJE4)).  So visualizers should be able to do this.  However, when I try to call drawingContext from the visualizer, I get the following error:
```
ERROR in .../frontscope/src/visualizers/VisualizerFactors.ts(294,39):
294:39 Property 'drawingContext' does not exist on type 'import(".../frontscope/node_modules/@types/p5/index.d.ts")'.
    292 | 
    293 |     grad_rect(x: number, y: number, width: number, height: number, color1: p5.Color, color2: p5.Color) {
  > 294 |         let barGradient = this.sketch.drawingContext.createLinearGradient(x, y, x, y+height);
        |                                       ^
    295 |         barGradient.addColorStop(0, color1);
    296 |         barGradient.addColorStop(1, color2);
    297 |         this.sketch.drawingContext.fillStyle = barGradient;
ERROR in .../frontscope/src/visualizers/VisualizerFactors.ts(297,21):
297:21 Property 'drawingContext' does not exist on type 'import("...frontscope/node_modules/@types/p5/index.d.ts")'.
    295 |         barGradient.addColorStop(0, color1);
    296 |         barGradient.addColorStop(1, color2);
  > 297 |         this.sketch.drawingContext.fillStyle = barGradient;
        |                     ^
    298 |         this.sketch.rect(x,y,width,height);
    299 |     }
    300 | }
```
Liam and I tried updating p5 to 1.4.0 but that didn't help.  

You can also circumvent this by getting the context from the canvas itself, but you need the HTML id in order to do this.  [Here's an example of how to do that](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createLinearGradient).  The problem here is that the visualizer doesn't know the canvas element id (in the browser it looks like 'DefaultCanvas0' or something like that when I inspect it), or I don't know the variable name it is stored in, if any.  When p5 creates a canvas using [createCanvas](https://p5js.org/reference/#/p5/createCanvas), it is possible to [set an id](https://p5js.org/reference/#/p5.Element/id).  But this seems to happen in ```src/components/CanvasArea.vue``` and ```src\components\bundleCard.vue``` and I don't know how to pass it from there back to the visualizer.

So, in short, the visualizers should have access to native HTML5 canvas context functionality, and the above gives some ideas how one might start figuring out how to do that.

**Comment by khaledallen** [comment]
My thought is that `canvasArea.vue` calls methods of the `activeViz`, so when you set the canvas id via `id(elementId)`, `canvasArea.vue` can pass the `elementId` used to the `activeViz`, or simply set it when instantiating. 

So [here](https://github.com/numberscope/frontscope/blob/16bdeb7770bb52773de70034d5c1f9c47ccb8f18/src/components/CanvasArea.vue#L48), hold the created canvas in some variable, create an id string, set it with `canvas.id(elementIdString)` and then somewhere below you can set `activeViz.setCanvasId(elementIdString)`.

Then the `activeViz` could write to it directly. If I recall correctly, this would "simply" require adding a new property to the vizualizer interface (something like `canvasElementId`) which can then be written to by anything that can control a vizualizer.

I also came across one SO thread with a similar issue: https://stackoverflow.com/questions/67348174/property-noise-does-not-exist-on-type-typeof-types-p5-index. Even the accepted solution speculates it may not be a clean solution, so I'd try the above solution first.

**Comment by katestange** [comment]
Apprently there's a notion of parentElement: ```const parentDiv = sketch.canvas.parentElement.id;``` (see [here](https://discourse.processing.org/t/access-properties-of-a-sketchs-container-in-instance-mode/10088/2) and [here](https://discourse.processing.org/t/how-to-get-atributes-from-sketch-its-parent-element/11452))

**Comment by liam-mulhall** [comment]
Part of @katestange's work on https://github.com/numberscope/frontscope/issues/115.

**Comment by gwhitney** [comment]
Try it again based on current main, the problem may have fixed itself. The very stupid code below replacing the usual `draw` method of VisualizerDifferences.ts just worked for me and drew a visible gradient at the top left of the visualizer area both as a bundle card and in the full-screen view:
```
    draw() {
        console.log(
            'Hi',
            this.sketch,
            this.sketch.noCanvas,
            this.sketch.drawingContext
        )
        this.drawDifferences(this.n, this.levels, this.seq)
        const bg = this.sketch.drawingContext.createLinearGradient(
            0,
            0,
            0,
            100
        )
        bg.addColorStop(0, 'white')
        bg.addColorStop(1, 'black')
        this.sketch.drawingContext.fillStyle = bg
        this.sketch.drawingContext.fillRect(0, 0, 50, 100)
        this.sketch.noLoop()
    }
```

**Comment by katestange** [comment]
This has fixed itself, yay!  Thanks Glen.  I'm going to close this issue.
